### DISCUSSION TITLE
How to use `COPY` statement with asyncpg dialect?

### DISCUSSION THREAD

**Discussion by MinsungKim-BlockCrafters** [sqlalchemy/sqlalchemy#8448] [Usage Questions] [answered]
Following code will print an error;

```python3
import asyncio
import io

from sqlalchemy.ext.asyncio import create_async_engine


async def main():
    url = "postgresql+asyncpg://minsungkim@localhost/test"
    engine = create_async_engine(url)
    buffer = io.BytesIO(b"hello,world\n")
    raw_conn = await engine.raw_connection()
    try:
        raw_conn.copy_to_table("test_table_name", source=buffer)
    except Exception as err:
        print(err)
    raw_conn.close()


if __name__ == "__main__":
    asyncio.run(main())
```

```
'AsyncAdapt_asyncpg_connection' object has no attribute 'copy_to_table'
```

I tried to implement code similar to [this stackoverflow question](https://stackoverflow.com/questions/13125236/sqlalchemy-psycopg2-and-postgresql-copy), but this doesn't work. Is there a way to do `COPY` without initializing new engine directly by `asyncpg`?

UPDATE: I am aware of executing raw sql but in such case I can't pass `io.BytesIO` buffer into COPY statement.

**Comment by zzzeek** [comment, marked as answer]
see https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncConnection.get_raw_connection

> Return the pooled DBAPI-level connection in use by this [AsyncConnection](https://docs.sqlalchemy.org/en/14/orm/extensions/asyncio.html#sqlalchemy.ext.asyncio.AsyncConnection).

>  This is a SQLAlchemy connection-pool proxied connection which then has the attribute [_ConnectionFairy.driver_connection](https://docs.sqlalchemy.org/en/14/core/pooling.html#sqlalchemy.pool._ConnectionFairy.driver_connection) that refers to the actual driver connection. Its [_ConnectionFairy.dbapi_connection](https://docs.sqlalchemy.org/en/14/core/pooling.html#sqlalchemy.pool._ConnectionFairy.dbapi_connection) refers instead to an [AdaptedConnection](https://docs.sqlalchemy.org/en/14/core/internals.html#sqlalchemy.engine.AdaptedConnection) instance that adapts the driver connection to the DBAPI protocol.

therefore use raw_connection().driver_connection to access asyncpg methods

**Reply by MinsungKim-BlockCrafters** [reply]
`(await Connection.get_raw_connection()).driver_connection` worked! Thank you!
